### PR TITLE
Enable brainpool curves for TLS1.3

### PR DESCRIPTION
**Pull request by bernd-edlinger**
See the recently assigned brainpool code points at:
https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml

### CODE REVIEW

File: ssl/ssl_locl.h
Comment (kroeckx) [inline ssl/ssl_locl.h]: Maybe we want to add defines for things like x25519 too, and then update eccurves_default[] to use the defines instead of just the numbers.

Comment (bernd-edlinger) [reply ssl/ssl_locl.h]: yes. there are two code points for P-256 and P-384

```
include/openssl/tls1.h:# define TLSEXT_curve_P_256                              23
include/openssl/tls1.h:# define TLSEXT_curve_P_384                              24
```

I was unsure why those are in a public API. Those have to stay, but
I would replace them with internal TLSEXT_GROUPID defines as needed.
Should I do that in this PR?

File: ssl/statem/statem_lib.c
Comment (kroeckx) [inline ssl/statem/statem_lib.c]: I really don't like this. I think you really should just add this like you would add a normal curve.

I also suggest that we add a minimum / maximum version for curves, I think that should simplify things.

**Comment by kroeckx** [comment]
tls1_set_groups() will need to be adjusted to support numbers >= 32.

**Comment by bernd-edlinger** [comment]
I hope I can avoid that.
The idea is I add NID_brainpoolP384r1 to set_groups and that gets translated
to a TLS13 groupid and a TLS12 based on the enabled protocols.
I would not like to add NID_brainpoolP384r1_for_tls13 or something.

**Comment by kroeckx** [comment]
Oh, right, you want to avoid adding NIDs.

**Comment by bernd-edlinger** [comment]
I just added a few macros to convert the group id in a cleaner way, untested so far.

**Comment by richsalz** [comment]
Also, please note that the Brainpool curves are explicitly *not in the recommended list.*  The registry requirement is just to have a specification.  I think that adding support for this goes against the stated OpenSSL policy, but the OMC will have to decide.

**Comment by kroeckx** [comment]
They should not be enabled by default, which this PR doesn't do as far as I know. You could argue that since TLS 1.3 says it shouldn't be supported, that they should do more than just enable it, they should enable it for use in TLS 1.3. And if you want that, you would need the new NIDs.

**Comment by bernd-edlinger** [comment]
It is definitely not intended to enable those curves by default.
The reason why I do this is as follows:
The OPC/UA Security Working Group https://opcfoundation.org/about/working-groups/
where I am a member of, is currently defining new OPC/UA ECC Security Profiles
using the following curves: nistP256, nistP384, X25519, X448, brainpoolP256r1, brainpoolP384r1.
Those profiles will require OPC/UA Servers to have certificates with those curves.
I just want to avoid to loose the ability to communicate using TLS in case such a certificate
is used.

**Comment by mattcaswell** [comment]
There isn't (yet) an RFC for this. The codepoints are there, but the associated RFC isn't. It seems a bit premature to add this.

**Comment by kaduk** [comment]
> There isn't (yet) an RFC for this. The codepoints are there, but the associated RFC isn't. It seems a bit premature to add this.

The recent changes to TLS registries (e.g., RFC 8447) explicitly intended to admit the possibility for codepoint assignment for Internet-Drafts that do not become RFCs; I don't think there's a priori a reason to wait for RFC publication.  The idea is that if the draft changes enough to not be interoperable then there would need to be new codepoints assigned.

**Comment by bernd-edlinger** [comment]
There is one thing that appears to be wrong.  I mean
If I use a P-256 certificate, and start the server with -curves brainpoolP256r1 or any
other curve like -curves secp256k1 (not P-256)
and connect the client with the same curves, then this tls1.2 code point is advertised
by the client, and accepted by the server, but still tls1.3 is negotiated.
This PR does only fix the behaviour for brainpool, but not for the ~20 other curves
without code point assignments.  Is this conforming to RFC 8446?

**Comment by kaduk** [comment]
I think the behavior you describe of negotiation 1.3 with a curve that is not defined for use in 1.3, is nonconformant to the spec.

**Review by paulidale — APPROVED** [review]
LGTM for master, although I'm not too conversant with this part of the code.

I'm not sure this is good for 3.0.  @openssl/otc ?

**Comment by mspncp** [comment]
> I'm not sure this is good for 3.0. @openssl/otc ?

I would consider it a new feature, not eligible for 3.0.

**Comment by t8m** [comment]
I do not see how this could be a bug fix. Clearly it is a new feature.

**Comment by beldmit** [comment]
+1 @t8m

**Comment by paulidale** [comment]
Dropping 3.0 tag.

**Comment by bernd-edlinger** [comment]
Merged to master as 0a10825.
Thanks!

**Comment by Basskrapfen** [comment]
Hi, we were hoping this change would be part of OpenSSL 3.1.x. 
Is it still possible to integrate it in OpenSSL 3.1.x?

Thanks a lot!

**Comment by Basskrapfen** [comment]
In case we would provide a backport to 3.1, would it be possible to merge this change into the 3.1 branch?

**Comment by t8m** [comment]
It would be considered based on the invasiveness of the change. If the patch appears soon though as the beta release which must be feature complete is now planned to be released in two or three weeks.

**Comment by Basskrapfen** [comment]
Thanks a lot, we will have a look how a backport could look like and how invasive it is.

**Comment by Basskrapfen** [comment]
Hi @t8m, I've tried and I think I'm able to apply the original patch in 3.1.x branch without many conflicts. If I insert the patch at the beginning of the 3.1.x branch history and reapply all changes from then, there is only one small conflict. I will try to create a clean pull request beginning of next week.

**Comment by t8m** [comment]
The problem is the original patch is not acceptable. It was later severally redone by @mattcaswell.
See: 
16f0e91cf82e13c327f0b0402459dfbf78ef787c
c9ee6e3646258f79a9970be96394cb2b93b7eddd
3f76339a3fe7be412b754ffe7b0a5438a1297f28

**Comment by t8m** [comment]
There is also a problem that OpenSSL 3.1 beta1 is currently scheduled for Thursday next week. Which would close the opportunity to add this support - actually it would have to be merged on Monday at the latest. I am afraid that is not doable.

Of course the schedule for 3.1 beta is not definite, it might be delayed if OTC decides so.

**Comment by Basskrapfen** [comment]
> There is also a problem that OpenSSL 3.1 beta1 is currently scheduled for Thursday next week. Which would close the opportunity to add this support - actually it would have to be merged on Monday at the latest. I am afraid that is not doable.
> 
> Of course the schedule for 3.1 beta is not definite, it might be delayed if OTC (on Tuesday) decides so.

Hi @t8m, the backport of all the patches was quite straight forward. We were able to apply them without many changes.
I've created a new pull request which is waiting for approval for the test. The final patch is luckily not very large. Please have a look, we would be very happy if it could make it into openssl-3.1.

If anything is missing, please let me know.

Thanks a lot!

**Pull request closed without merging**
